File: sdk-kotlin/SynapseSDK.kt
/**
 * SYNAPSE Protocol Kotlin SDK
 * 
 * For Android applications and Kotlin backend services
 * 
 * Implementation:
 *   implementation("ai.synapse:synapse-sdk:1.0.0")
 */

package ai.synapse.sdk

import kotlinx.coroutines.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import org.web3j.abi.FunctionEncoder
import org.web3j.abi.FunctionReturnDecoder
import org.web3j.abi.TypeReference
import org.web3j.abi.datatypes.*
import org.web3j.abi.datatypes.generated.Uint256
import org.web3j.crypto.Credentials
import org.web3j.crypto.RawTransaction
import org.web3j.crypto.TransactionEncoder
import org.web3j.protocol.Web3j
import org.web3j.protocol.core.DefaultBlockParameterName
import org.web3j.protocol.http.HttpService
import org.web3j.tx.gas.DefaultGasProvider
import org.web3j.utils.Convert
import org.web3j.utils.Numeric
import java.math.BigDecimal
import java.math.BigInteger

/**
 * Configuration for the SDK
 */
data class SynapseConfig(
    val rpcUrl: String = "https://arb1.arbitrum.io/rpc",
    val chainId: Long = 42161,
    val tokenAddress: String = "",
    val routerAddress: String = "",
    val stakingAddress: String = "",
    val apiBaseUrl: String = "https://api.synapse.ai"
)

/**
 * Transaction result
 */
data class TransactionResult(
    val hash: String,
    val from: String,
    val to: String,
    val value: BigInteger,
    val success: Boolean,
    val blockNumber: Long? = null,
    val gasUsed: BigInteger? = null
)

/**
 * Stake information
 */
data class StakeInfo(
    val amount: BigDecimal,
    val lockTier: Int,
    val lockEnd: Long,
    val pendingRewards: BigDecimal
) {
    val isLocked: Boolean get() = lockEnd > System.currentTimeMillis() / 1000
    
    fun getFormattedAmount(): String = "$amount SYNX"
    fun getFormattedRewards(): String = "$pendingRewards SYNX"
}

/**
 * Payment details
 */
data class PaymentInfo(
    val paymentId: String,
    val sender: String,
    val recipient: String,
    val amount: BigDecimal,
    val timestamp: Long,
    val status: PaymentStatus,
    val metadata: String? = null
)

enum class PaymentStatus {
    PENDING, COMPLETED, FAILED, REFUNDED
}

/**
 * Escrow details
 */
data class EscrowInfo(
    val escrowId: String,
    val sender: String,
    val recipient: String,
    val arbiter: String?,
    val amount: BigDecimal,
    val deadline: Long,
    val status: EscrowStatus
)

enum class EscrowStatus {
    ACTIVE, RELEASED, REFUNDED, DISPUTED, EXPIRED
}

/**
 * Main SDK Client
 */
class SynapseClient(
    private val config: SynapseConfig
) {
    private val web3j: Web3j = Web3j.build(HttpService(config.rpcUrl))
    private val httpClient = OkHttpClient.Builder().build()
    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()
    
    private var credentials: Credentials? = null
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    /**
     * Set signer with private key
     */
    fun setSigner(privateKey: String): SynapseClient {
        credentials = Credentials.create(privateKey)
        return this
    }

    /**
     * Get connected address
     */
    fun getAddress(): String? = credentials?.address

    // ============ Token Operations ============

    /**
     * Get token balance
     */
    suspend fun getBalance(address: String? = null): BigDecimal = withContext(Dispatchers.IO) {
        val targetAddress = address ?: credentials?.address 
            ?: throw IllegalStateException("No address provided")

        val function = Function(
            "balanceOf",
            listOf(Address(targetAddress)),
            listOf(object : TypeReference<Uint256>() {})
        )

        val encodedFunction = FunctionEncoder.encode(function)
        val response = web3j.ethCall(
            org.web3j.protocol.core.methods.request.Transaction.createEthCallTransaction(
                targetAddress, config.tokenAddress, encodedFunction
            ),
            DefaultBlockParameterName.LATEST
        ).send()

        val results = FunctionReturnDecoder.decode(response.value, function.outputParameters)
        val balance = results[0] as Uint256

        Convert.fromWei(BigDecimal(balance.value), Convert.Unit.ETHER)
    }

    /**
     * Transfer tokens
     */
    suspend fun transfer(to: String, amount: BigDecimal): TransactionResult = withContext(Dispatchers.IO) {
        requireSigner()

        val amountWei = Convert.toWei(amount, Convert.Unit.ETHER).toBigInteger()

        val function = Function(
            "transfer",
            listOf(Address(to), Uint256(amountWei)),
            listOf(object : TypeReference<Bool>() {})
        )

        sendTransaction(config.tokenAddress, BigInteger.ZERO, FunctionEncoder.encode(function))
    }

    /**
     * Approve token spending
     */
    suspend fun approve(spender: String, amount: BigDecimal): TransactionResult = withContext(Dispatchers.IO) {
        requireSigner()

        val amountWei = Convert.toWei(amount, Convert.Unit.ETHER).toBigInteger()

        val function = Function(
            "approve",
            listOf(Address(spender), Uint256(amountWei)),
            emptyList()
        )

        sendTransaction(config.tokenAddress, BigInteger.ZERO, FunctionEncoder.encode(function))
    }

    /**
     * Get allowance
     */
    suspend fun getAllowance(owner: String, spender: String): BigDecimal = withContext(Dispatchers.IO) {
        val function = Function(
            "allowance",
            listOf(Address(owner), Address(spender)),
            listOf(object : TypeReference<Uint256>() {})
        )

        val encodedFunction = FunctionEncoder.encode(function)
        val response = web3j.ethCall(
            org.web3j.protocol.core.methods.request.Transaction.createEthCallTransaction(
                owner, config.tokenAddress, encodedFunction
            ),
            DefaultBlockParameterName.LATEST
        ).send()

        val results = FunctionReturnDecoder.decode(response.value, function.outputParameters)
        val allowance = results[0] as Uint256

        Convert.fromWei(BigDecimal(allowance.value), Convert.Unit.ETHER)
    }

    // ============ Payment Operations ============

    /**
     * Send payment
     */
    suspend fun sendPayment(
        recipient: String,
        amount: BigDecimal,
        metadata: String = ""
    ): TransactionResult = withContext(Dispatchers.IO) {
        requireSigner()

        // Ensure approval
        ensureApproval(config.routerAddress, amount)

        val amountWei = Convert.toWei(amount, Convert.Unit.ETHER).toBigInteger()

        val function = Function(
            "pay",
            listOf(
                Address(recipient),
                Uint256(amountWei),
                Utf8String(metadata)
            ),
            emptyList()
        )

        sendTransaction(config.routerAddress, BigInteger.ZERO, FunctionEncoder.encode(function))
    }

    /**
     * Create escrow payment
     */
    suspend fun createEscrow(
        recipient: String,
        amount: BigDecimal,
        deadline: Long,
        arbiter: String? = null
    ): TransactionResult = withContext(Dispatchers.IO) {
        requireSigner()

        ensureApproval(config.routerAddress, amount)

        val amountWei = Convert.toWei(amount, Convert.Unit.ETHER).toBigInteger()
        val arbiterAddress = arbiter ?: "0x0000000000000000000000000000000000000000"

        val function = Function(
            "createEscrow",
            listOf(
                Address(recipient),
                Address(arbiterAddress),
                Uint256(amountWei),
                Uint256(BigInteger.valueOf(deadline))
            ),
            emptyList()
        )

        sendTransaction(config.routerAddress, BigInteger.ZERO, FunctionEncoder.encode(function))
    }

    /**
     * Release escrow
     */
    suspend fun releaseEscrow(escrowId: String): TransactionResult = withContext(Dispatchers.IO) {
        requireSigner()

        val function = Function(
            "releaseEscrow",
            listOf(Bytes32(Numeric.hexStringToByteArray(escrowId))),
            emptyList()
        )

        sendTransaction(config.routerAddress, BigInteger.ZERO, FunctionEncoder.encode(function))
    }

    // ============ Staking Operations ============

    /**
     * Stake tokens
     */
    suspend fun stake(amount: BigDecimal, lockTier: Int = 0): TransactionResult = withContext(Dispatchers.IO) {
        requireSigner()

        ensureApproval(config.stakingAddress, amount)

        val amountWei = Convert.toWei(amount, Convert.Unit.ETHER).toBigInteger()

        val function = Function(
            "stake",
            listOf(Uint256(amountWei), Uint256(BigInteger.valueOf(lockTier.toLong()))),
            emptyList()
        )

        sendTransaction(config.stakingAddress, BigInteger.ZERO, FunctionEncoder.encode(function))
    }

    /**
     * Unstake tokens
     */
    suspend fun unstake(amount: BigDecimal): TransactionResult = withContext(Dispatchers.IO) {
        requireSigner()

        val amountWei = Convert.toWei(amount, Convert.Unit.ETHER).toBigInteger()

        val function = Function(
            "unstake",
            listOf(Uint256(amountWei)),
            emptyList()
        )

        sendTransaction(config.stakingAddress, BigInteger.ZERO, FunctionEncoder.encode(function))
    }

    /**
     * Claim staking rewards
     */
    suspend fun claimRewards(): TransactionResult = withContext(Dispatchers.IO) {
        requireSigner()

        val function = Function(
            "claimRewards",
            emptyList(),
            emptyList()
        )

        sendTransaction(config.stakingAddress, BigInteger.ZERO, FunctionEncoder.encode(function))
    }

    /**
     * Get stake info
     */
    suspend fun getStakeInfo(address: String? = null): StakeInfo = withContext(Dispatchers.IO) {
        val targetAddress = address ?: credentials?.address
            ?: throw IllegalStateException("No address provided")

        val function = Function(
            "getStakeInfo",
            listOf(Address(targetAddress)),
            listOf(
                object : TypeReference<Uint256>() {},
                object : TypeReference<Uint256>() {},
                object : TypeReference<Uint256>() {},
                object : TypeReference<Uint256>() {}
            )
        )

        val encodedFunction = FunctionEncoder.encode(function)
        val response = web3j.ethCall(
            org.web3j.protocol.core.methods.request.Transaction.createEthCallTransaction(
                targetAddress, config.stakingAddress, encodedFunction
            ),
            DefaultBlockParameterName.LATEST
        ).send()

        val results = FunctionReturnDecoder.decode(response.value, function.outputParameters)

        StakeInfo(
            amount = Convert.fromWei(BigDecimal((results[0] as Uint256).value), Convert.Unit.ETHER),
            lockTier = (results[1] as Uint256).value.toInt(),
            lockEnd = (results[2] as Uint256).value.toLong(),
            pendingRewards = Convert.fromWei(BigDecimal((results[3] as Uint256).value), Convert.Unit.ETHER)
        )
    }

    /**
     * Get pending rewards
     */
    suspend fun getPendingRewards(address: String? = null): BigDecimal = withContext(Dispatchers.IO) {
        val targetAddress = address ?: credentials?.address
            ?: throw IllegalStateException("No address provided")

        val function = Function(
            "earned",
            listOf(Address(targetAddress)),
            listOf(object : TypeReference<Uint256>() {})
        )

        val encodedFunction = FunctionEncoder.encode(function)
        val response = web3j.ethCall(
            org.web3j.protocol.core.methods.request.Transaction.createEthCallTransaction(
                targetAddress, config.stakingAddress, encodedFunction
            ),
            DefaultBlockParameterName.LATEST
        ).send()

        val results = FunctionReturnDecoder.decode(response.value, function.outputParameters)
        Convert.fromWei(BigDecimal((results[0] as Uint256).value), Convert.Unit.ETHER)
    }

    // ============ API Calls ============

    /**
     * Get payment history from API
     */
    suspend fun getPaymentHistory(address: String? = null, limit: Int = 50): List<PaymentInfo> = 
        withContext(Dispatchers.IO) {
            val targetAddress = address ?: credentials?.address
                ?: throw IllegalStateException("No address provided")

            val request = Request.Builder()
                .url("${config.apiBaseUrl}/api/payments/$targetAddress?limit=$limit")
                .get()
                .build()

            val response = httpClient.newCall(request).execute()
            val body = response.body?.string() ?: return@withContext emptyList()

            val json = JSONObject(body)
            val payments = json.getJSONArray("payments")

            (0 until payments.length()).map { i ->
                val p = payments.getJSONObject(i)
                PaymentInfo(
                    paymentId = p.getString("paymentId"),
                    sender = p.getString("sender"),
                    recipient = p.getString("recipient"),
                    amount = BigDecimal(p.getString("amount")),
                    timestamp = p.getLong("timestamp"),
                    status = PaymentStatus.valueOf(p.getString("status").uppercase()),
                    metadata = p.optString("metadata")
                )
            }
        }

    /**
     * Get price from oracle
     */
    suspend fun getTokenPrice(): BigDecimal = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("${config.apiBaseUrl}/api/price/synx")
            .get()
            .build()

        val response = httpClient.newCall(request).execute()
        val body = response.body?.string() ?: throw Exception("Empty response")

        val json = JSONObject(body)
        BigDecimal(json.getString("price"))
    }

    // ============ Internal Functions ============

    private fun requireSigner() {
        if (credentials == null) {
            throw IllegalStateException("No signer set. Call setSigner() first.")
        }
    }

    private suspend fun ensureApproval(spender: String, amount: BigDecimal) {
        val currentAllowance = getAllowance(credentials!!.address, spender)
        if (currentAllowance < amount) {
            // Approve max
            approve(spender, BigDecimal("115792089237316195423570985008687907853269984665640564039457584007913129639935"))
        }
    }

    private suspend fun sendTransaction(
        to: String,
        value: BigInteger,
        data: String
    ): TransactionResult = withContext(Dispatchers.IO) {
        val nonce = web3j.ethGetTransactionCount(
            credentials!!.address,
            DefaultBlockParameterName.PENDING
        ).send().transactionCount

        val gasPrice = web3j.ethGasPrice().send().gasPrice
        val gasLimit = estimateGas(to, value, data)

        val rawTransaction = RawTransaction.createTransaction(
            nonce,
            gasPrice,
            gasLimit,
            to,
            value,
            data
        )

        val signedMessage = TransactionEncoder.signMessage(rawTransaction, config.chainId, credentials)
        val hexValue = Numeric.toHexString(signedMessage)

        val transactionHash = web3j.ethSendRawTransaction(hexValue).send().transactionHash

        TransactionResult(
            hash = transactionHash,
            from = credentials!!.address,
            to = to,
            value = value,
            success = true
        )
    }

    private suspend fun estimateGas(to: String, value: BigInteger, data: String): BigInteger {
        return try {
            val estimate = web3j.ethEstimateGas(
                org.web3j.protocol.core.methods.request.Transaction.createFunctionCallTransaction(
                    credentials!!.address,
                    null,
                    null,
                    null,
                    to,
                    value,
                    data
                )
            ).send()

            // Add 20% buffer
            estimate.amountUsed.multiply(BigInteger.valueOf(120)).divide(BigInteger.valueOf(100))
        } catch (e: Exception) {
            DefaultGasProvider.GAS_LIMIT
        }
    }

    /**
     * Cleanup resources
     */
    fun shutdown() {
        scope.cancel()
        web3j.shutdown()
    }
}

/**
 * Builder for SynapseClient
 */
class SynapseClientBuilder {
    private var config = SynapseConfig()
    private var privateKey: String? = null

    fun rpcUrl(url: String) = apply { config = config.copy(rpcUrl = url) }
    fun chainId(id: Long) = apply { config = config.copy(chainId = id) }
    fun tokenAddress(address: String) = apply { config = config.copy(tokenAddress = address) }
    fun routerAddress(address: String) = apply { config = config.copy(routerAddress = address) }
    fun stakingAddress(address: String) = apply { config = config.copy(stakingAddress = address) }
    fun apiBaseUrl(url: String) = apply { config = config.copy(apiBaseUrl = url) }
    fun privateKey(key: String) = apply { privateKey = key }

    fun build(): SynapseClient {
        val client = SynapseClient(config)
        privateKey?.let { client.setSigner(it) }
        return client
    }
}

// Extension functions for easy usage
fun synapseClient(block: SynapseClientBuilder.() -> Unit): SynapseClient {
    return SynapseClientBuilder().apply(block).build()
}

File: sdk-php/SynapseSDK.php
<?php
/**
 * SYNAPSE Protocol PHP SDK
 * 
 * For Laravel, Symfony, and other PHP applications
 * 
 * Installation: composer require synapse/sdk
 */

declare(strict_types=1);

namespace Synapse\SDK;

use Web3\Web3;
use Web3\Contract;
use Web3\Utils;
use kornrunner\Keccak;
use kornrunner\Ethereum\Transaction;

/**
 * Main SDK Client
 */
class SynapseClient
{
    private Web3 $web3;
    private array $contracts = [];
    private array $config;
    private ?string $privateKey = null;
    private ?string $address = null;

    /**
     * Constructor
     * 
     * @param array $config Configuration options
     */
    public function __construct(array $config)
    {
        $this->config = array_merge([
            'rpcUrl' => 'https://arb1.arbitrum.io/rpc',
            'chainId' => 42161,
            'tokenAddress' => '',
            'routerAddress' => '',
            'stakingAddress' => '',
        ], $config);

        $this->web3 = new Web3($this->config['rpcUrl']);
    }

    /**
     * Set signer
     */
    public function setSigner(string $privateKey): self
    {
        $this->privateKey = $privateKey;
        $this->address = $this->privateKeyToAddress($privateKey);
        return $this;
    }

    /**
     * Get connected address
     */
    public function getAddress(): ?string
    {
        return $this->address;
    }

    // ============ Token Operations ============

    /**
     * Get token balance
     */
    public function getBalance(string $address = null): string
    {
        $address = $address ?? $this->address;
        if (!$address) {
            throw new \Exception('No address provided');
        }

        $contract = $this->getTokenContract();
        $balance = '0';

        $contract->call('balanceOf', $address, function ($err, $result) use (&$balance) {
            if ($err) throw new \Exception($err->getMessage());
            $balance = $result[0]->toString();
        });

        return $balance;
    }

    /**
     * Get formatted balance
     */
    public function getFormattedBalance(string $address = null): string
    {
        $balance = $this->getBalance($address);
        return Utils::fromWei($balance, 'ether');
    }

    /**
     * Transfer tokens
     */
    public function transfer(string $to, string $amount): TransactionResult
    {
        $this->requireSigner();
        
        $contract = $this->getTokenContract();
        $data = $contract->getData('transfer', $to, Utils::toWei($amount, 'ether'));
        
        return $this->sendTransaction($this->config['tokenAddress'], '0', $data);
    }

    /**
     * Approve spending
     */
    public function approve(string $spender, string $amount): TransactionResult
    {
        $this->requireSigner();

        $contract = $this->getTokenContract();
        $data = $contract->getData('approve', $spender, Utils::toWei($amount, 'ether'));

        return $this->sendTransaction($this->config['tokenAddress'], '0', $data);
    }

    /**
     * Get allowance
     */
    public function getAllowance(string $owner, string $spender): string
    {
        $contract = $this->getTokenContract();
        $allowance = '0';

        $contract->call('allowance', $owner, $spender, function ($err, $result) use (&$allowance) {
            if ($err) throw new \Exception($err->getMessage());
            $allowance = $result[0]->toString();
        });

        return $allowance;
    }

    // ============ Payment Operations ============

    /**
     * Send payment
     */
    public function sendPayment(string $recipient, string $amount, string $metadata = ''): TransactionResult
    {
        $this->requireSigner();

        // Ensure approval
        $this->ensureApproval($this->config['routerAddress'], $amount);

        $contract = $this->getRouterContract();
        $data = $contract->getData('pay', $recipient, Utils::toWei($amount, 'ether'), $metadata);

        return $this->sendTransaction($this->config['routerAddress'], '0', $data);
    }

    /**
     * Create escrow
     */
    public function createEscrow(
        string $recipient,
        string $amount,
        int $deadline,
        string $arbiter = null
    ): TransactionResult {
        $this->requireSigner();
        
        $this->ensureApproval($this->config['routerAddress'], $amount);

        $contract = $this->getRouterContract();
        $data = $contract->getData(
            'createEscrow',
            $recipient,
            $arbiter ?? '0x0000000000000000000000000000000000000000',
            Utils::toWei($amount, 'ether'),
            $deadline
        );

        return $this->sendTransaction($this->config['routerAddress'], '0', $data);
    }

    /**
     * Release escrow
     */
    public function releaseEscrow(string $escrowId): TransactionResult
    {
        $this->requireSigner();

        $contract = $this->getRouterContract();
        $data = $contract->getData('releaseEscrow', $escrowId);

        return $this->sendTransaction($this->config['routerAddress'], '0', $data);
    }

    // ============ Staking Operations ============

    /**
     * Stake tokens
     */
    public function stake(string $amount, int $lockTier = 0): TransactionResult
    {
        $this->requireSigner();

        $this->ensureApproval($this->config['stakingAddress'], $amount);

        $contract = $this->getStakingContract();
        $data = $contract->getData('stake', Utils::toWei($amount, 'ether'), $lockTier);

        return $this->sendTransaction($this->config['stakingAddress'], '0', $data);
    }

    /**
     * Unstake tokens
     */
    public function unstake(string $amount): TransactionResult
    {
        $this->requireSigner();

        $contract = $this->getStakingContract();
        $data = $contract->getData('unstake', Utils::toWei($amount, 'ether'));

        return $this->sendTransaction($this->config['stakingAddress'], '0', $data);
    }

    /**
     * Claim rewards
     */
    public function claimRewards(): TransactionResult
    {
        $this->requireSigner();

        $contract = $this->getStakingContract();
        $data = $contract->getData('claimRewards');

        return $this->sendTransaction($this->config['stakingAddress'], '0', $data);
    }

    /**
     * Get stake info
     */
    public function getStakeInfo(string $address = null): StakeInfo
    {
        $address = $address ?? $this->address;
        if (!$address) {
            throw new \Exception('No address provided');
        }

        $contract = $this->getStakingContract();
        $info = null;

        $contract->call('getStakeInfo', $address, function ($err, $result) use (&$info) {
            if ($err) throw new \Exception($err->getMessage());
            $info = $result;
        });

        return new StakeInfo([
            'amount' => $info['amount']->toString(),
            'lockTier' => (int)$info['lockTier']->toString(),
            'lockEnd' => (int)$info['lockEnd']->toString(),
            'pendingRewards' => $info['pendingRewards']->toString(),
        ]);
    }

    /**
     * Get pending rewards
     */
    public function getPendingRewards(string $address = null): string
    {
        $address = $address ?? $this->address;
        
        $contract = $this->getStakingContract();
        $rewards = '0';

        $contract->call('earned', $address, function ($err, $result) use (&$rewards) {
            if ($err) throw new \Exception($err->getMessage());
            $rewards = $result[0]->toString();
        });

        return $rewards;
    }

    // ============ Internal Methods ============

    private function requireSigner(): void
    {
        if (!$this->privateKey || !$this->address) {
            throw new \Exception('No signer set. Call setSigner() first.');
        }
    }

    private function ensureApproval(string $spender, string $amount): void
    {
        $amountWei = Utils::toWei($amount, 'ether');
        $currentAllowance = $this->getAllowance($this->address, $spender);

        if (gmp_cmp($currentAllowance, $amountWei) < 0) {
            $maxApproval = '0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff';
            $this->approve($spender, Utils::fromWei($maxApproval, 'ether'));
        }
    }

    private function sendTransaction(string $to, string $value, string $data): TransactionResult
    {
        $nonce = $this->getNonce();
        $gasPrice = $this->getGasPrice();
        $gasLimit = $this->estimateGas($to, $value, $data);

        $tx = new Transaction(
            $nonce,
            $gasPrice,
            $gasLimit,
            $to,
            $value,
            $data
        );

        $signedTx = $tx->sign($this->privateKey, $this->config['chainId']);

        $txHash = null;
        $this->web3->eth->sendRawTransaction('0x' . $signedTx, function ($err, $hash) use (&$txHash) {
            if ($err) throw new \Exception($err->getMessage());
            $txHash = $hash;
        });

        return new TransactionResult([
            'hash' => $txHash,
            'from' => $this->address,
            'to' => $to,
            'value' => $value,
        ]);
    }

    private function getNonce(): string
    {
        $nonce = '0';
        $this->web3->eth->getTransactionCount($this->address, 'pending', function ($err, $count) use (&$nonce) {
            if ($err) throw new \Exception($err->getMessage());
            $nonce = '0x' . dechex((int)$count->toString());
        });
        return $nonce;
    }

    private function getGasPrice(): string
    {
        $gasPrice = '0';
        $this->web3->eth->gasPrice(function ($err, $price) use (&$gasPrice) {
            if ($err) throw new \Exception($err->getMessage());
            $gasPrice = '0x' . dechex((int)$price->toString());
        });
        return $gasPrice;
    }

    private function estimateGas(string $to, string $value, string $data): string
    {
        $gas = '0x55f0'; // Default 22000
        $this->web3->eth->estimateGas([
            'from' => $this->address,
            'to' => $to,
            'value' => $value,
            'data' => $data,
        ], function ($err, $estimated) use (&$gas) {
            if (!$err) {
                $gas = '0x' . dechex((int)((int)$estimated->toString() * 1.2)); // Add 20% buffer
            }
        });
        return $gas;
    }

    private function privateKeyToAddress(string $privateKey): string
    {
        $privateKey = ltrim($privateKey, '0x');
        $publicKey = secp256k1_ec_pubkey_create($privateKey);
        $hash = Keccak::hash(hex2bin($publicKey), 256);
        return '0x' . substr($hash, -40);
    }

    private function getTokenContract(): Contract
    {
        if (!isset($this->contracts['token'])) {
            $this->contracts['token'] = new Contract($this->web3->provider, ABIs::TOKEN);
            $this->contracts['token']->at($this->config['tokenAddress']);
        }
        return $this->contracts['token'];
    }

    private function getRouterContract(): Contract
    {
        if (!isset($this->contracts['router'])) {
            $this->contracts['router'] = new Contract($this->web3->provider, ABIs::PAYMENT_ROUTER);
            $this->contracts['router']->at($this->config['routerAddress']);
        }
        return $this->contracts['router'];
    }

    private function getStakingContract(): Contract
    {
        if (!isset($this->contracts['staking'])) {
            $this->contracts['staking'] = new Contract($this->web3->provider, ABIs::STAKING);
            $this->contracts['staking']->at($this->config['stakingAddress']);
        }
        return $this->contracts['staking'];
    }
}

/**
 * Transaction Result
 */
class TransactionResult
{
    public string $hash;
    public string $from;
    public string $to;
    public string $value;
    public ?array $receipt = null;

    public function __construct(array $data)
    {
        $this->hash = $data['hash'];
        $this->from = $data['from'];
        $this->to = $data['to'];
        $this->value = $data['value'];
    }

    public function toArray(): array
    {
        return [
            'hash' => $this->hash,
            'from' => $this->from,
            'to' => $this->to,
            'value' => $this->value,
            'receipt' => $this->receipt,
        ];
    }
}

/**
 * Stake Info
 */
class StakeInfo
{
    public string $amount;
    public int $lockTier;
    public int $lockEnd;
    public string $pendingRewards;

    public function __construct(array $data)
    {
        $this->amount = $data['amount'];
        $this->lockTier = $data['lockTier'];
        $this->lockEnd = $data['lockEnd'];
        $this->pendingRewards = $data['pendingRewards'];
    }

    public function getFormattedAmount(): string
    {
        return Utils::fromWei($this->amount, 'ether');
    }

    public function getFormattedRewards(): string
    {
        return Utils::fromWei($this->pendingRewards, 'ether');
    }

    public function isLocked(): bool
    {
        return $this->lockEnd > time();
    }

    public function toArray(): array
    {
        return [
            'amount' => $this->amount,
            'formattedAmount' => $this->getFormattedAmount(),
            'lockTier' => $this->lockTier,
            'lockEnd' => $this->lockEnd,
            'pendingRewards' => $this->pendingRewards,
            'formattedRewards' => $this->getFormattedRewards(),
            'isLocked' => $this->isLocked(),
        ];
    }
}

/**
 * ABIs
 */
class ABIs
{
    public const TOKEN = '[
        {"constant":true,"inputs":[{"name":"account","type":"address"}],"name":"balanceOf","outputs":[{"name":"","type":"uint256"}],"type":"function"},
        {"constant":false,"inputs":[{"name":"to","type":"address"},{"name":"amount","type":"uint256"}],"name":"transfer","outputs":[{"name":"","type":"bool"}],"type":"function"},
        {"constant":false,"inputs":[{"name":"spender","type":"address"},{"name":"amount","type":"uint256"}],"name":"approve","outputs":[{"name":"","type":"bool"}],"type":"function"},
        {"constant":true,"inputs":[{"name":"owner","type":"address"},{"name":"spender","type":"address"}],"name":"allowance","outputs":[{"name":"","type":"uint256"}],"type":"function"}
    ]';

    public const PAYMENT_ROUTER = '[
        {"inputs":[{"name":"recipient","type":"address"},{"name":"amount","type":"uint256"},{"name":"metadata","type":"string"}],"name":"pay","outputs":[{"name":"","type":"bytes32"}],"type":"function"},
        {"inputs":[{"name":"recipient","type":"address"},{"name":"arbiter","type":"address"},{"name":"amount","type":"uint256"},{"name":"deadline","type":"uint256"}],"name":"createEscrow","outputs":[{"name":"","type":"bytes32"}],"type":"function"},
        {"inputs":[{"name":"escrowId","type":"bytes32"}],"name":"releaseEscrow","outputs":[],"type":"function"}
    ]';

    public const STAKING = '[
        {"inputs":[{"name":"amount","type":"uint256"},{"name":"lockTier","type":"uint8"}],"name":"stake","outputs":[],"type":"function"},
        {"inputs":[{"name":"amount","type":"uint256"}],"name":"unstake","outputs":[],"type":"function"},
        {"inputs":[],"name":"claimRewards","outputs":[],"type":"function"},
        {"inputs":[{"name":"user","type":"address"}],"name":"getStakeInfo","outputs":[{"components":[{"name":"amount","type":"uint256"},{"name":"lockTier","type":"uint8"},{"name":"lockEnd","type":"uint256"},{"name":"pendingRewards","type":"uint256"},{"name":"lastClaim","type":"uint256"}],"name":"","type":"tuple"}],"type":"function"},
        {"inputs":[{"name":"user","type":"address"}],"name":"earned","outputs":[{"name":"","type":"uint256"}],"type":"function"}
    ]';
}
